fabric-tool-attributes-v1/src/main/java/net/fabricmc/fabric/mixin/tools/MixinItemStack.java

```diff
@@ -42,12 +51,27 @@ public void isEffectiveOn(BlockState state, CallbackInfoReturnable<Boolean> info
 	}
 
 	@Inject(at = @At("HEAD"), method = "getMiningSpeed", cancellable = true)
-	public void getBlockBreakingSpeed(BlockState state, CallbackInfoReturnable<Float> info) {
-		if (this.getItem() instanceof MiningToolItemAccessor) {
-			TriState triState = ToolManager.handleIsEffectiveOn((ItemStack) (Object) this, state);
-			if (triState != TriState.DEFAULT) {
-				info.setReturnValue(triState.get() ? ((MiningToolItemAccessor) this.getItem()).getMiningSpeed() : 1.0F);
-				info.cancel();
+	public void getMiningSpeed(BlockState state, CallbackInfoReturnable<Float> info) {
+		TriState triState = ToolManager.handleIsEffectiveOn((ItemStack) (Object) this, state);
+		if (triState != TriState.DEFAULT) {
+			Item item = this.getItem();
+			float miningSpeed;
+			if (item instanceof ToolAttributeHolder) {
+				miningSpeed = ((ToolAttributeHolder) this.getItem()).getMiningSpeed((ItemStack)(Object) this);
+			} else {
+				return;
+			}
+			info.setReturnValue(triState.get() ?  miningSpeed : 1.0F);
+		}
+	}
+
+	@Inject(at = @At("RETURN"), method = "getAttributeModifiers", cancellable = true, locals = LocalCapture.CAPTURE_FAILEXCEPTION)
+	public void getAttributeModifiers(EquipmentSlot slot, CallbackInfoReturnable<Multimap<String, EntityAttributeModifier>> info, Multimap<String, EntityAttributeModifier> multimap) {
+		ItemStack stack = (ItemStack) (Object) this;
+		if (stack.getItem() instanceof ToolAttributeHolder) {
+			if (!stack.hasTag() || !stack.getTag().containsKey("fabric_IgnoreDynamicModifiers")) {
```

Comment:
I think we should discuss how we want to approach this. Is it really Fabric's responsibility to add this kind of game logic, or should perhaps implementations of ToolAttributeHolder do it?

Reply:
You can put this logic into the interface as a default method that modders can override is they wish.

Reply:
Yeah, I think I'm gonna add a `shouldUseDynamicTags` method.